04-onion-routing.md

```diff
@@ -76,81 +69,59 @@ The use of a fixed nonce is safe since the keys are never reused.
 
 ## Packet Structure
 
-The packet consists of 2 parts:
+The packet consists of 4 parts:
  
- - The fixed size _header_ containing meta information about the
-   packet and the routing information necessary to forward the
-   message.
- - A fixed size _per-hop payload_ containing information for each hop
+ - A `version` byte
+ - A 33-byte compressed `secp256k1` `public key`, used during the shared secret generation
+ - A 1300-byte `hops_data` consisting of twenty fixed size packets containing information for each hop
    as they forward the message.
+ - A 32-byte `HMAC` used to verify the packet's integrity.
 
 The overall structure of the packet is depicted below. The network format of the packet consists of the individual parts being serialized into one continguous byte-stream and then transferred to the recipient of the packet. Due to the fixed size of the packet it does not need to be prefixed by its length when transferred over a connection.
 
 ~~~~
-+--------+-----------------+
-| header | per-hop payload |
-+--------+-----------------+
++------------------+-----------------------+------------------------+-----------------+
+| version (1 byte) | public key (33 bytes) | hops_data (20x65 bytes) | HMAC (32 bytes) |
++------------------+-----------------------+------------------------+-----------------+
```

Comment:
This is just the HMAC extracted by hop `i` form `hops_data` and passed to hop `i+1` so that it can check the integrity. So it's not a separate HMAC, but one of the `hops_data` HMACs decrypted and exposed.

This also means that the HMACs are shifted by one, leaving the last HMAC blank, which is what we currently use to identify the terminal. We could rework it so that every hop decrypts its own HMAC, but that may be a bigger rework in the handling logic.